apps/labs/posts/uarray-intro.md

```diff
@@ -0,0 +1,331 @@
+        111 
+        112         return result
+
+    BackendNotImplementedError: No selected backends had an implementation for this method.
+
+Now we are told that no backends had an implementation for this function
+(which is nice, good error messages are nice!)
+
+## Coercion and passing between backends
+
+Let's say we had two `Backend`s. Let's choose the completely useless
+example of one storing a number as an `int` and one as a `float`.
+
+``` python
+class Number(ua.DispatchableInstance):
+    pass
+
+def myfunc_rd(args, kwargs, dispatchable_args):
+    # Here, we're "replacing" the dispatchable args with the ones supplied.
+    # In general, this may be more complex, like inserting them in between
+    # other args and kwargs.
+    return dispatchable_args, kwargs
+
+@ua.create_multimethod(myfunc_rd)
+def myfunc(a):
+    # Here, we're marking a as a Number, and saying that "we want to dispatch/convert over this"
+    # We return as a tuple as there may be more dispatchable arguments
+    return (Number(a),)
+
+Number.register_convertor(be1, lambda x: int(x))
+Number.register_convertor(be2, lambda x: str(x))
+```
+
+Let's also define a \"catch-all\" method. This catches all
+implementations of methods not already registered.
+
+``` python
+# This can be arbitrarily complex
+def gen_impl1(method, args, kwargs, dispatchable_args):
+    if not all(isinstance(a, Number) and isinstance(a.value, int) for a in dispatchable_args):
+        return NotImplemented
+    
+    return args[0]
+
+# This can be arbitrarily complex
+def gen_impl2(method, args, kwargs, dispatchable_args):
+    if not all(isinstance(a, Number) and isinstance(a.value, str) for a in dispatchable_args):
+        return NotImplemented
+    
+    return args[0]
+
+be1.register_implementation(None, gen_impl1)
+be2.register_implementation(None, gen_impl2)
+```
+
+``` python
+myfunc('1') # This calls the second implementation
+```
+
+    '1'
+
+``` python
+myfunc(1) # This calls the first implementation
+```
+
+    1
+
+``` python
+myfunc(1.0) # This fails
+```
+
+    ---------------------------------------------------------------------------
+    BackendNotImplementedError                Traceback (most recent call last)
+    <ipython-input-13-8431c1275db5> in <module>
+    ----> 1 myfunc(1.0) # This fails
+
+    ~/Quansight/uarray/uarray/backend.py in __call__(self, *args, **kwargs)
+        108 
+        109         if result is NotImplemented:
+    --> 110             raise BackendNotImplementedError('No selected backends had an implementation for this method.')
+        111 
+        112         return result
+
+    BackendNotImplementedError: No selected backends had an implementation for this method.
+
+``` python
+# But works if we do this:
+
+with ua.set_backend(be1, coerce=True):
+    print(type(myfunc(1.0)))
+
+with ua.set_backend(be2, coerce=True):
+    print(type(myfunc(1.0)))
+```
+
+    <class 'int'>
```

Comment:
@gabalafou I'm getting the Vercel build error: `error [Error: [next-mdx-remote] error compiling MDX:`

Full error message
```
Unexpected character `'` (U+0027) after name,
expected a character that can start an attribute name,
such as a letter, `$`, or `_`; whitespace before attributes;
or the end of the tag

  270 |
> 271 |     <class 'int'>
      |            ^
  272 |     <class 'str'>
  273 |
```

[Link to error](https://vercel.com/quansight/labs/7bDxZz8gpDGNVX5y1VZmSHTeUNBd#:~:text=15%3A02%3A27.909-,error%20%5BError%3A%20%5Bnext%2Dmdx%2Dremote%5D%20error%20compiling%20MDX%3A,-15%3A02%3A27.910)

The context is showing the result of some code. [Link to this section in the old blog website.](https://labs.quansight.org/blog/2019/04/uarray-intro#:~:text=%3C-,class,-%27int%27%3E%0A%3Cclass%20%27str)

Maybe we could wrap it in a code block? But then it would look out of place with the rest of the code results, unless the rest were also in a code block.

Reply:
Ugh that is formatted so ugly on the old site. If I understand correctly, and as I think you're saying, on this blog post on the old website, the blocks with gray background are meant to represent the code blocks, whereas the monospace-font lines (with no background color) that immediately follow are meant to represent the output from running those code blocks. But the way that is shown visually in the old blog post isn't intuitive to me at all. Let me play with this a bit and get back to you. Thanks for providing the link to the old blog post, by the way, super helpful!

Reply:
Thanks for taking a look at it! And yes, I understood the blog the same way, code and then result. Maybe both could be in code blocks, but have different color backgrounds or text, to emphasize the difference?